# Laboratorio 2 — Números (pseudo)aleatorios y Generación de Variables Aleatorias
## Sección 5.1 — Generación de números (pseudo)aleatorios

**Curso:** Simulación Computacional
**Guía:** Práctica N° 02  
**Docente:** Angel Alfonso Cruz Roa  
**Estudiante:** Luciana A. Belandria A.  
**Código:** *160005005*



## 0. Librerías y dependencias

In [2]:

import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import pandas as pd
from scipy import stats
from collections import Counter

np.set_printoptions(precision=4, suppress=True)

N = 100
ALPHA = 0.05

SECUENCIAS = {}

def showValues(I, X, U):
    print("i\tXi\tUi")
    for t in range(0, len(X)):
        print("%d\t%d\t%2.4f" % (I[t], X[t], U[t]), sep=' ', end='\n')

## 5.1.1 Generador de cuadrados medios (MidSquare)

Tomando en cuenta el tema visto en clase en las diapositivas de Números Aleatorios implemente o use un
generador con el método de MidSquare o Middle-Square (o cuadrados medios) para generar una secuencia ui
= u1, u2, …, un de n =100 números pseudoaleatorios usando las siguientes semillas:


In [3]:
def midSquare(xant, d):

    cuadrado = str(xant**2).zfill(2*d)
    ini = (2*d - d)//2
    xsig = int(cuadrado[ini:ini+d])
    usig = xsig / (10**d)
    return [xsig, usig]

def midSquareN(x0, N):

    d = len(str(x0))
    xant = x0
    I, X, U = [], [], []
    for t in range(1, N+1):
        [xi, ui] = midSquare(xant, d)
        I.append(t); X.append(xi); U.append(ui)
        xant = xi
    return [I, X, U]

In [5]:
semillas_ms = [1931, 675248, 4567, 12122, 7179345]

for x0 in semillas_ms:
    d = len(str(x0))
    [I, X, U] = midSquareN(x0, N)
    nombre = "MidSquare X0=%d" % x0
    SECUENCIAS[nombre] = {
        'U': U, 'X': X,
        'estado0': x0,
        'siguiente': (lambda d: (lambda x: midSquare(x, d)[0]))(d)
    }
    print("="*70)
    print("%s   (d = %d dígitos)" % (nombre, d))
    print("Primeros 10 valores:")
    showValues(I[:10], X[:10], U[:10])
    print("u_i (primeros 15):", np.round(U[:15], 4))

MidSquare X0=1931   (d = 4 dígitos)
Primeros 10 valores:
i	Xi	Ui
1	7287	0.7287
2	1003	0.1003
3	60	0.0060
4	36	0.0036
5	12	0.0012
6	1	0.0001
7	0	0.0000
8	0	0.0000
9	0	0.0000
10	0	0.0000
u_i (primeros 15): [0.7287 0.1003 0.006  0.0036 0.0012 0.0001 0.     0.     0.     0.
 0.     0.     0.     0.     0.    ]
MidSquare X0=675248   (d = 6 dígitos)
Primeros 10 valores:
i	Xi	Ui
1	959861	0.9599
2	333139	0.3331
3	981593	0.9816
4	524817	0.5248
5	432883	0.4329
6	387691	0.3877
7	304311	0.3043
8	605184	0.6052
9	247673	0.2477
10	341914	0.3419
u_i (primeros 15): [0.9599 0.3331 0.9816 0.5248 0.4329 0.3877 0.3043 0.6052 0.2477 0.3419
 0.9052 0.3563 0.9233 0.5291 0.8992]
MidSquare X0=4567   (d = 4 dígitos)
Primeros 10 valores:
i	Xi	Ui
1	8574	0.8574
2	5134	0.5134
3	3579	0.3579
4	8092	0.8092
5	4804	0.4804
6	784	0.0784
7	6146	0.6146
8	7733	0.7733
9	7992	0.7992
10	8720	0.8720
u_i (primeros 15): [0.8574 0.5134 0.3579 0.8092 0.4804 0.0784 0.6146 0.7733 0.7992 0.872
 0.0384 0.1474 0.1726 0.979  0.8441]
MidSqu

## 5.1.2 Generador congruencial mixto (GLC)

Tomando en cuenta el tema visto en clase en las diapositivas de Números Aleatorios y el Capítulo 3 Números
Aleatorios de [Ross99], implemente o use el método de generador congruencial mixto o generador lineal
congruencial (GLC) de acuerdo con la siguiente ecuación:

$$X_{n+1} = (a X_n + c) \bmod m, \qquad u_i = \frac{X_i}{m}$$
Para generar una secuencia ui = u1, u2, …, un de n=100 números pseudoaleatorios


In [6]:
def genran(a, c, m, xant):
    xsig = (a*xant + c) % m
    usig = xsig / m
    return [xsig, usig]

def genranN(a, c, m, x0, N):
    xant = x0
    I = []
    X = []
    U = []
    for t in range(1, N+1):
        [xi, ui] = genran(a, c, m, xant)
        I.append(t)
        X.append(xi)
        U.append(ui)
        xant = xi
    return [I, X, U]

In [7]:
# (etiqueta, X0, a, c, m)
params_glc = [
    ('a', 127,      115,         0,  128),
    ('b', 43,       25,          5,  48),
    ('c', 13,       16807,       0,  2**31 - 1),
    ('d', 29022024, 25214903917, 11, 2**48),
    ('e', 3,        5,           7,  200),
]

for etq, x0, a, c, m in params_glc:
    [I, X, U] = genranN(a, c, m, x0, N)
    nombre = "GLC(%s) X0=%d,a=%d,c=%d,m=%d" % (etq, x0, a, c, m)
    SECUENCIAS[nombre] = {
        'U': U, 'X': X,
        'estado0': x0,
        'siguiente': (lambda a, c, m: (lambda x: (a*x + c) % m))(a, c, m)
    }
    print("="*70)
    print(nombre)
    print("Primeros 10 valores:")
    showValues(I[:10], X[:10], U[:10])
    print("u_i (primeros 15):", np.round(U[:15], 4))

GLC(a) X0=127,a=115,c=0,m=128
Primeros 10 valores:
i	Xi	Ui
1	13	0.1016
2	87	0.6797
3	21	0.1641
4	111	0.8672
5	93	0.7266
6	71	0.5547
7	101	0.7891
8	95	0.7422
9	45	0.3516
10	55	0.4297
u_i (primeros 15): [0.1016 0.6797 0.1641 0.8672 0.7266 0.5547 0.7891 0.7422 0.3516 0.4297
 0.4141 0.6172 0.9766 0.3047 0.0391]
GLC(b) X0=43,a=25,c=5,m=48
Primeros 10 valores:
i	Xi	Ui
1	24	0.5000
2	29	0.6042
3	10	0.2083
4	15	0.3125
5	44	0.9167
6	1	0.0208
7	30	0.6250
8	35	0.7292
9	16	0.3333
10	21	0.4375
u_i (primeros 15): [0.5    0.6042 0.2083 0.3125 0.9167 0.0208 0.625  0.7292 0.3333 0.4375
 0.0417 0.1458 0.75   0.8542 0.4583]
GLC(c) X0=13,a=16807,c=0,m=2147483647
Primeros 10 valores:
i	Xi	Ui
1	218491	0.0001
2	1524694590	0.7100
3	1767098126	0.8229
4	2066849319	0.9625
5	1988514208	0.9260
6	1817779242	0.8465
7	1313358072	0.6116
8	1772192238	0.8252
9	1784243823	0.8309
10	324286453	0.1510
u_i (primeros 15): [0.0001 0.71   0.8229 0.9625 0.926  0.8465 0.6116 0.8252 0.8309 0.151
 0.9855 0.7524 0.8025 0.4494 0.695 ]

## 5.1.3 Otros generadores de números aleatorios

Tomando en cuenta la lectura del artículo científico [Mancilla2000], implemente al menos dos (2) métodos de los
siguientes generadores de numeros pseudoaleatorios y genere una secuencia ui = u1, u2, …, un de n =100
números aleatorios con unos parámetros de ejemplo que Ud. escoja.


### a. Generador Shift-Register

$$X_{n+k} = \left(\sum_{i=0}^{k-1} a_i X_{n+i}\right) \bmod 2$$

donde los $X_n$ y los $a_i$ son 0 o 1, y el período máximo es $2^{k}-1$. Se usa la realimentación de dos términos $x_n = (x_{n-k} + x_{n-q}) \bmod 2$ con $k = 31$ y $q = 3$ (polinomio primitivo $x^{31}+x^{3}+1$), en la variante **multi-step**: se toman $L = 31$ bits sucesivos para formar cada entero.

### b. Generador Lagged-Fibonacci

$$X_n = (X_{n-j} \oplus X_{n-k}) \bmod M, \quad j < k, \; M = 2^{m}$$

con el operador $\oplus$ = suma, retardos $j = 5$, $k = 17$ y $M = 2^{31}$. La tabla inicial de $k$ valores se llena con un GLC.



In [15]:
# a. Shift-Register
K_SR, Q_SR, L_SR = 31, 3, 31

def paso_shift_register(estado):

    bits = list(estado)
    for _ in range(L_SR):
        nuevo = bits[0] ^ bits[K_SR - Q_SR]
        bits.pop(0)
        bits.append(nuevo)
    return tuple(bits)

def genShiftRegisterN(estado0, N):
    estado = tuple(estado0)
    X, U = [], []
    for _ in range(N):
        estado = paso_shift_register(estado)
        val = 0
        for b in estado[-L_SR:]:
            val = (val << 1) | b
        X.append(val)
        U.append(val / (2**L_SR))
    return X, U


estado0_sr = tuple(int(b) for b in format(29022024, '031b'))
X_sr, U_sr = genShiftRegisterN(estado0_sr, N)

SECUENCIAS["Shift-Register k=31,q=3"] = {
    'U': U_sr, 'X': X_sr, 'estado0': estado0_sr, 'siguiente': paso_shift_register
}
print("Shift-Register -> u_i (primeros 15):", np.round(U_sr[:15], 4))

Shift-Register -> u_i (primeros 15): [0.0121 0.7095 0.2025 0.2088 0.8005 0.9482 0.4888 0.549  0.2659 0.6964
 0.0637 0.0705 0.7945 0.2618 0.9866]


In [17]:
# ---------- b. Lagged-Fibonacci ----------
J_LF, K_LF, M_LF = 5, 17, 2**31

def paso_lagged_fibonacci(estado):
    """estado = tupla con los últimos K_LF valores (el más antiguo primero)."""
    tabla = list(estado)
    nuevo = (tabla[K_LF - J_LF] + tabla[0]) % M_LF   # X_{n-j} + X_{n-k}
    tabla.pop(0)
    tabla.append(nuevo)
    return tuple(tabla)

def genLaggedFibonacciN(estado0, N):
    estado = tuple(estado0)
    X, U = [], []
    for _ in range(N):
        estado = paso_lagged_fibonacci(estado)
        X.append(estado[-1])
        U.append(estado[-1] / M_LF)
    return X, U


[_, X_ini, _] = genranN(1103515245, 12345, M_LF, 29022024, K_LF)
estado0_lf = tuple(X_ini)
X_lf, U_lf = genLaggedFibonacciN(estado0_lf, N)

SECUENCIAS["Lagged-Fibonacci j=5,k=17"] = {
    'U': U_lf, 'X': X_lf, 'estado0': estado0_lf, 'siguiente': paso_lagged_fibonacci
}
print("Lagged-Fibonacci -> u_i (primeros 15):", np.round(U_lf[:15], 4))

Lagged-Fibonacci -> u_i (primeros 15): [0.243  0.293  0.5508 0.2168 0.6863 0.8203 0.3323 0.2453 0.5666 0.536
 0.3359 0.1616 0.5833 0.4021 0.6777]


## 5.1.4 Tamaño de ciclo, pruebas de uniformidad y aleatoriedad

Para cada una de las secuencias generadas en 5.1.1, 5.1.2 y 5.1.3 se calcula:

a. Tamaño de ciclo o periodo de la secuencia si lo tiene.

b. Prueba de uniformidad con alguno de los métodos vistos en clase

c. Prueba de aleatoriedad con el método visto en clase

De acuerdo con el tamaño de ciclo encontrado y las pruebas de uniformidad y aleatoriedad, responda:

• ¿Cuáles son los peores generadores de números aleatorios y porqué?

• ¿Cuáles son los mejores generadores de números aleatorios y porqué?

### (a) Tamaño de ciclo o período




In [13]:
def detectar_ciclo(estado0, siguiente, max_iter=200000):

    vistos = {}
    estado = estado0
    for i in range(max_iter):
        if estado in vistos:
            mu = vistos[estado]
            lam = i - mu
            return mu, lam
        vistos[estado] = i
        estado = siguiente(estado)
    return None, None

# tope de iteraciones
TOPES = {"Shift-Register k=31,q=3": 20000}

periodos = {}
for nombre, s in SECUENCIAS.items():
    tope = TOPES.get(nombre, 200000)
    mu, lam = detectar_ciclo(s['estado0'], s['siguiente'], max_iter=tope)
    periodos[nombre] = (mu, lam, tope)
    if lam is None:
        print("%-34s  período > %d (no se detecta dentro del tope)" % (nombre, tope))
    else:
        deg = "sí (se degenera)" if mu > 0 else "no"
        print("%-34s  período λ = %-8d transitorio μ = %-6d ¿degenera?: %s" % (nombre, lam, mu, deg))

MidSquare X0=1931                   período λ = 1        transitorio μ = 7      ¿degenera?: sí (se degenera)
MidSquare X0=675248                 período λ = 1        transitorio μ = 210    ¿degenera?: sí (se degenera)
MidSquare X0=4567                   período λ = 4        transitorio μ = 19     ¿degenera?: sí (se degenera)
MidSquare X0=12122                  período λ = 1        transitorio μ = 88     ¿degenera?: sí (se degenera)
MidSquare X0=7179345                período λ = 20       transitorio μ = 848    ¿degenera?: sí (se degenera)
GLC(a) X0=127,a=115,c=0,m=128       período λ = 32       transitorio μ = 0      ¿degenera?: no
GLC(b) X0=43,a=25,c=5,m=48          período λ = 48       transitorio μ = 0      ¿degenera?: no
GLC(c) X0=13,a=16807,c=0,m=2147483647  período > 200000 (no se detecta dentro del tope)
GLC(d) X0=29022024,a=25214903917,c=11,m=281474976710656  período > 200000 (no se detecta dentro del tope)
GLC(e) X0=3,a=5,c=7,m=200           período λ = 8        transitorio μ 

### (b) Prueba de uniformidad 1: contraste $\chi^2$

Se divide $(0,1)$ en $k$ clases de igual amplitud; con $f_i$ observaciones en la clase $i$ y $e_i = n/k$ esperadas bajo $H_0$ (uniformidad),

$$\chi^{2}=\sum_{i=1}^{k}\frac{(f_{i}-e_{i})^{2}}{e_{i}} \sim \chi^{2}_{k-1}$$

Se rechaza la hipótesis de uniformidad si $\chi^{2} > \chi^{2}_{k-1,\alpha}$. Con $n = 100$ se toma $k = 10$, de modo que $e_i = 10 > 5$ y la aproximación asintótica es válida.

In [19]:
def test_chi2_uniformidad(U, k=10, alpha=ALPHA):
    n = len(U)
    f, _ = np.histogram(U, bins=k, range=(0.0, 1.0))
    e = n/k
    T = np.sum((f - e)**2 / e)
    Tval = stats.chi2.ppf(1 - alpha, k - 1)
    pvalue = 1 - stats.chi2.cdf(T, k - 1)
    return T, Tval, pvalue, (T > Tval)

nombre_demo = list(SECUENCIAS.keys())[0]
T, Tval, pv, rechaza = test_chi2_uniformidad(SECUENCIAS[nombre_demo]['U'])
print("Secuencia:", nombre_demo)
print("Chi^2 = %.4f   Chi^2_{k-1,alpha} = %.4f   p-value = %.4f" % (T, Tval, pv))
print("Se rechaza la hipotesis de uniformidad." if rechaza else "Se acepta la hipotesis de uniformidad.")

Secuencia: MidSquare X0=1931
Chi^2 = 860.6000   Chi^2_{k-1,alpha} = 16.9190   p-value = 0.0000
Se rechaza la hipotesis de uniformidad.


### (c) Prueba de aleatoriedad: test de rachas

Dada la muestra $u_1,\dots,u_n$ se construye una sucesión de 0 y 1: en la posición $i$ se coloca 0 si $u_{i+1} < u_i$ y 1 si $u_{i+1} > u_i$. Cada grupo consecutivo de 0's o de 1's es una **racha**. Para $n$ grande el número de rachas $R$ sigue

$$R \sim \mathcal{N}\left(\frac{2n-1}{3},\; \sigma^{2}=\frac{16n-29}{90}\right), \qquad Z = \frac{R-\frac{2n-1}{3}}{\sqrt{\frac{16n-29}{90}}}\sim\mathcal{N}(0,1)$$

Se rechaza la hipótesis de aleatoriedad si $|Z| > Z_{\alpha/2}$.

>

In [20]:
def test_rachas(U, alpha=ALPHA):
    n = len(U)
    A = np.zeros(n-1)
    for i in range(1, n):
        if U[i-1] > U[i]:
            A[i-1] = 0
        else:
            A[i-1] = 1

    A2 = np.array((A == 1))
    sw = (A2[:-1] ^ A2[1:])
    isw = np.arange(len(sw))[sw]
    isw = list(isw); isw.insert(0, -1)
    isw = np.array(isw)
    L = isw[1:len(isw)] - isw[0:len(isw)-1]

    R = len(L) + 1
    Z = (R - ((2*n - 1)/3)) / (np.sqrt((16*n - 29)/90))
    Zt = stats.norm.ppf((1 + (1 - alpha))/2)
    return R, Z, Zt, L, (abs(Z) > Zt)

R, Z, Zt, L, rechaza = test_rachas(SECUENCIAS[nombre_demo]['U'])
print("Secuencia:", nombre_demo)
print("Longitudes de las primeras rachas:", L[:20])
print("R = %d   Z = %.4f   Z_{alpha/2} = %.4f" % (R, Z, Zt))
print("Se rechaza la hipotesis de aleatoriedad." if rechaza else "Se acepta la hipotesis de aleatoriedad.")

Secuencia: MidSquare X0=1931
Longitudes de las primeras rachas: [6]
R = 2   Z = -15.3982   Z_{alpha/2} = 1.9600
Se rechaza la hipotesis de aleatoriedad.


### Respuestas a las preguntas

De acuerdo con el tamaño de ciclo encontrado y las pruebas de uniformidad y aleatoriedad, responda:

• ¿Cuáles son los peores generadores de números aleatorios y porqué?

• ¿Cuáles son los mejores generadores de números aleatorios y porqué?

#### ¿Cuáles son los peores generadores de números aleatorios y por qué?

**1. MidSquare con semilla 1931 .** La sucesión se degenera en apenas 7 iteraciones y cae en el punto fijo $X=0$, de modo que $\lambda = 1$: de los 100 números generados solo hay 7 valores distintos y el resto son ceros. Rechaza las cuatro pruebas ($\chi^2 = 860.6$, $D = 0.974$, $T_{pares} = 382.4$, $Z = -15.4$).

**2. MidSquare con semilla 4567.** Transitorio $\mu = 19$ y período $\lambda = 4$: 22 valores distintos en 100. Rechaza las cuatro pruebas.

**3. MidSquare con semilla 12122.** También colapsa en un punto fijo ($\lambda = 1$) tras 88 pasos y rechaza uniformidad y aleatoriedad.

**4. GLC(e): $X_0=3$, $a=5$, $c=7$, $m=200$.** Se degenera ($\mu = 2$, $\lambda = 8$): solo 9 valores distintos. La causa se explica con las condiciones de [Mancilla2000] para período máximo: $m = 200 = 2^3\cdot 5^2$, y aunque $c=7$ es primo relativo con $m$, $a-1 = 4$ **no** es múltiplo de 5, que es factor primo de $m$. Rechaza $\chi^2$ y pares consecutivos.

**5. GLC(a): $X_0=127$, $a=115$, $c=0$, $m=128$ y GLC(b): $X_0=43$, $a=25$, $c=5$, $m=48$.** No se degeneran (GLC(b) alcanza incluso el período máximo $\lambda^*(48) = 48$, igual que la tabla 4 de [Mancilla2000]), pero su período es tan pequeño que en 100 números la secuencia completa se repite 2 o 3 veces. GLC(a), por ser multiplicativo ($c = 0$) con $m = 2^7$, solo alcanza $\lambda = 32 = m/4$, tal como predice la teoría del método multiplicativo. Aunque pasan $\chi^2$, KS y rachas —justamente porque recorren el ciclo completo y eso "parece" uniforme—, GLC(b) rechaza el test de pares consecutivos no solapados, que sí detecta la estructura repetitiva. Un generador con período corto es inservible para simulación aunque pase las pruebas de uniformidad.

**MidSquare con semillas 675248 y 7179345.** Pasan las cuatro pruebas, pero no son buenos generadores: su transitorio es $\mu = 210$ y $\mu = 848$ respectivamente, es decir, más largo que los 100 números pedidos, así que la degeneración simplemente todavía no se alcanza a ver. Sus períodos reales son $\lambda = 1$ y $\lambda = 20$. Esto ilustra por qué el tamaño de ciclo debe analizarse aparte de las pruebas estadísticas.

#### ¿Cuáles son los mejores generadores de números aleatorios y por qué?

**1. GLC(d): $X_0=29022024$, $a=25214903917$, $c=11$, $m=2^{48}$** (los parámetros de `drand48`). Es el mejor del conjunto: no se degenera, su período supera el tope de búsqueda de $2\times10^5$ (teóricamente $2^{48}\approx 2.8\times10^{14}$, porque cumple las condiciones de período máximo: $c$ impar primo relativo con $m$, y $a-1$ múltiplo de 4 y de 2) y pasa las cuatro pruebas con holgura ($\chi^2 = 4.8$, $D = 0.083$, $T_{pares} = 5.8$, $Z = -0.80$).

**2. Shift-Register y Lagged-Fibonacci.** Los cuatro generadores de la sección 5.1.3 pasan también las cuatro pruebas y tienen períodos enormes (para el combinado, del orden de $M_1 \cdot M_2 \approx 4.6\times10^{18}$; para el Shift-Register, $2^{31}-1$). El combinado y el de congruencia inversa son los mejores en los diagramas de dispersión $(u_i, u_{i+1})$, donde no se observa estructura reticular.

**3. GLC(c): $X_0=13$, $a=16807$, $c=0$, $m=2^{31}-1$** (el clásico MINSTD de Lehmer). Tiene período enorme y pasa $\chi^2$, pares consecutivos y rachas, pero **rechaza** el test de Kolmogorov-Smirnov por poco ($D = 0.1452$ frente a $D_{crit} = 0.1340$). Con solo $n = 100$ números y $\alpha = 0.05$ una prueba puede rechazar por azar aproximadamente 1 de cada 20 veces; lo correcto, como sugiere el notebook *Test para Números Aleatorios*, es **repetir el contraste** con otras semillas o con $n$ mayor antes de descartar el generador.

#### Criterios usados para clasificarlos

Siguiendo a [Mancilla2000], un buen generador debe cumplir: **período máximo**, **uniformidad**, **independencia / secuencias no correlacionadas**, **eficiencia** de cómputo y **reproducibilidad**. En este laboratorio el criterio que más discrimina no son las pruebas estadísticas sino el **tamaño de ciclo**: todos los generadores rechazados o dudosos tienen períodos ridículamente cortos o se degeneran, mientras que los generadores con módulo grande y parámetros que cumplen las condiciones de período máximo pasan todas las pruebas.


### Bibliografía

- [Ross99] Ross, Sheldon. *Simulación*, 2da Edición. Pearson Press, 1999. Capítulo 3.
- [Mancilla2000] Mancilla Herrera, Alfonso Manuel. *Números aleatorios. Historia, teoría y aplicaciones*. Ingeniería y Desarrollo, núm. 8, 2000, pp. 49-69. Universidad del Norte.
- [Ríos, Ríos & Martín, 2000] David Ríos, Sixto Ríos, Jacinto Martín. *Simulación, Métodos y aplicaciones*. Alfaomega, 2000.